### DISCUSSION TITLE
Broadcasting RTCM3 correction data as NTRIP server/caster

### DISCUSSION THREAD

**Discussion by HamzaCetin0** [semuconsulting/pygnssutils#70] [Q&A] [answered]
Hi, thank you for amazing work. I have a question about steaming RTCM data as NTRIP server. I want to stream the RTCM data as a NTRIP server and than some other users(clients) to get the data.  I use "gnssserver --inport "/dev/tty.ACM0" --hostip 0.0.0.0 --outport 2101 --ntripmode 1 --protfilter 4 --format 2 --ntripuser myuser --ntrippassword mypassword" command to stream the data. Then on another terminal I run "gnssdump --socket 0.0.0.0:2101" command to connect the localhost and get the RTCM data as a client. After connecting the network on the server side it it shows that I am connected but on the client side nothing happend. I cant see any data flow on the terminal. I dont know what it is wrong. Can you help me with this? 

![client](https://github.com/semuconsulting/pygnssutils/assets/161038558/959d29fa-08cb-4434-a6a7-cf4427a67159)
![server](https://github.com/semuconsulting/pygnssutils/assets/161038558/c913c866-4a98-4d09-aead-c34e382ca9b2)

**Comment by semuadmin** [comment, marked as answer]
Hi @HamzaCetin0,

If your intention is to run `gnssserver` as an NTRIP Caster, you will need to use an NTRIP Client such as [`gnssntripclient`](https://github.com/semuconsulting/pygnssutils?tab=readme-ov-file#gnssntripclient) to receive the data, rather than `gnssdump` which is basically just a serial parser.

So, based on your example above, you would need to use something like:

```shell
gnssntripclient --server localhost --mountpoint pygnssutils --ntripuser myuser --ntrippassword mypassword
```
You should then see a stream of incoming RTCM3 data something like this...

```
2024-06-12 11:09:58.030652: Using mountpoint pygnssutils

2024-06-12 11:09:58.034132: <RTCM(1077, DF002=1077, DF003=0, DF004=295816001, DF393=1, DF409=0, DF001_7=0, DF411=0, DF412=0, DF417=0, DF418=0, DF394=234547957876129792, NSat=7, ..., DF404_07=-0.39680000000000004)>
2024-06-12 11:09:58.034893: <RTCM(1087, DF002=1087, DF003=0, DF416=3, DF034=47398001, DF393=1, DF409=0, DF001_7=0, DF411=0, DF412=0, DF417=0, DF418=0, DF394=73751941456330752, NSat=5, ..., DF404_05=0.038700000000000005)>
...etc.
```

**NB:** Bear in mind that, for all this to work, the receiver at `/dev/tty.ACM0` _must_ be an RTK-compatible receiver (_e.g. ZED-F9P_) which is configured as an RTK base station - `gnssserver` will _not_ do this automatically. See https://github.com/semuconsulting/pyubx2/blob/master/examples/f9p_basestation.py for an illustration of how to do this using `pyubx2`, or you could use [PyGPSClient's NTRIP Caster functionality](https://github.com/semuconsulting/PyGPSClient?tab=readme-ov-file#socketserver).

I'm converting this Issue to a DIscussion Q&A topic - if you have any further queries, please add them to the discussion.

Hope this helps.

**Comment by HamzaCetin0** [comment]
Thank you so much for your reply. Now everything is ok. I can see the data on the terminal. I have one more question about the output of the data. Is there any way to get the data by typing the local address and the port number on the browser? And how to save this data to a file?

**Reply by semuadmin** [reply]
Hi again,

Although NTRIP is essentially an HTTP protocol, you would typically need to use an NTRIP Client of some kind to retrieve the data in a usable format. You could, at a push, access the data via a simple web browser, but it would be a fairly clunky process and may not work with some caster/browser combinations.

For example, if you have `gnssserver` running as per your example above and enter the URL `http://localhost:2101/pygnssutils` into a web browser, then (_depending on your browser's security configuration_) you'll first be prompted for your NTRIP login credentials:

<img width="600" alt="Screenshot 2024-06-12 at 13 16 04" src="https://github.com/semuconsulting/pygnssutils/assets/28569967/e600e4c4-5e26-4150-a5bf-fc981612f0fa">

Once you've entered valid credentials, then (_depending on your browser's MIME and Downloads configuration_) the ensuing RTCM data is treated as a generic binary data stream which can be saved as a binary file (_you may need to pause the download in order to access the saved *.part data file_):

<img width="600" alt="Screenshot 2024-06-12 at 13 16 29" src="https://github.com/semuconsulting/pygnssutils/assets/28569967/a4015ac8-fa7c-4365-975a-05b663e4caf6">

**Note that** if you simply want to access an RTCM correction data stream within your own private LAN or WAN, you don't _necessarily_ have to use NTRIP - you could simply stream the raw binary RTCM data via a TCP socket e.g.
```shell
gnssserver -I "/dev/tty.ACM0" -H 0.0.0.0 -O 51000 -N 0
```
... and consume it using any TCP socket reader e.g.
```shell
gnssdump -S localhost:51000
```

Again, you could access this data stream via a web browser, but it will simply appear as binary gibberish:

<img width="600" alt="Screenshot 2024-06-12 at 13 36 02" src="https://github.com/semuconsulting/pygnssutils/assets/28569967/4a56a681-e6c6-47c6-969d-3a9a13162646">


To save the _parsed_ RTCM data to a file using `gnssntripclient`, you can use the `--logtofile 1` option:
```shell
gnssntripclient --server localhost --mountpoint pygnssutils --ntripuser myuser --ntrippassword mypassword --verbosity 2 --logtofile 1
```

The `gnssntripclient` CLI utility doesn't currently support saving _binary_ RTCM3 data to a file (_I may add this functionality in a future release_), but it's relatively straightforward to write a Python script to do this using its underlying `GNSSNTRIPClient` class - see, for example: https://github.com/semuconsulting/pygnssutils/blob/main/examples/rtcm_ntrip_client.py.

**Reply by HamzaCetin0** [reply]
Got it. Thank you so much for your contribution.

**Reply by HamzaCetin0** [reply]
Hi again. I have a question about the mountpoint. When I changed in the mountpoint from "pygnssutils" to any other mountpoint I could not connect to the server and cant get any RTCM data.  Is there a way to  set  a custom mountpoint  to stream the RTCM data?
